core/src/telemetry/tracing.ts

```diff
@@ -0,0 +1,363 @@
+ *
+ * @param params The parameters object containing agent and invocation context.
+ *
+ * Inference related fields are not set, due to their planned removal from invoke_agent span:
+ * https://github.com/open-telemetry/semantic-conventions/issues/2632
+ *
+ * `gen_ai.agent.id` is not set because currently it's unclear what attributes this field should have, specifically:
+ * - In which scope should it be unique (globally, given project, given agentic flow, given deployment).
+ * - Should it be unchanging between deployments, and how this should this be achieved.
+ *
+ * `gen_ai.data_source.id` is not set because it's not available.
+ * Closest type which could contain this information is types.GroundingMetadata, which does not have an ID.
+ *
+ * `server.*` attributes are not set pending confirmation from aabmass.
+ */
+export function traceAgentInvocation({
+  agent,
+  invocationContext,
+}: TraceAgentInvocationParams): void {
+  const span = trace.getActiveSpan();
+  if (!span) return;
+  
+  // Required
+  span.setAttributes({
+    [GEN_AI_OPERATION_NAME]: 'invoke_agent',
+    // Conditionally Required
+    [GEN_AI_AGENT_DESCRIPTION]: agent.description || '',
+    [GEN_AI_AGENT_NAME]: agent.name,
+    [GEN_AI_CONVERSATION_ID]: invocationContext.session.id,
+  });
+}
+
+export interface TraceToolCallParams {
+  /** The tool that was called. */
+  tool: BaseTool;
+  /** The arguments to the tool call. */
+  args: Record<string, unknown>;
+  /** The event with the function response details. */
+  functionResponseEvent: Event;
+}
+
+/**
+ * Traces tool call.
+ *
+ * @param params The parameters object containing tool, args, and function response event.
+ */
+export function traceToolCall({
+  tool,
+  args,
+  functionResponseEvent,
+}: TraceToolCallParams): void {
+  const span = trace.getActiveSpan();
+  if (!span) return;
+  
+  span.setAttributes({
+    [GEN_AI_OPERATION_NAME]: 'execute_tool',
+    [GEN_AI_TOOL_DESCRIPTION]: tool.description || '',
+    [GEN_AI_TOOL_NAME]: tool.name,
+    // e.g. FunctionTool
+    [GEN_AI_TOOL_TYPE]: tool.constructor.name,
+    // Setting empty llm request and response (as UI expect these) while not
+    // applicable for tool_response.
+    'gcp.vertex.agent.llm_request': '{}',
+    'gcp.vertex.agent.llm_response': '{}',
+    'gcp.vertex.agent.tool_call_args': safeJsonSerialize(args),
+  });
+
+  // Tracing tool response
+  let toolCallId = '<not specified>';
+  let toolResponse: unknown = '<not specified>';
+  
+  if (functionResponseEvent.content?.parts) {
+    const responseParts = functionResponseEvent.content.parts;
+    const functionResponse = responseParts[0]?.functionResponse;
+    if (functionResponse) {
+      if (functionResponse.id) {
+        toolCallId = functionResponse.id;
+      }
+      if (functionResponse.response) {
+        toolResponse = functionResponse.response;
+      }
+    }
```

Comment:
can you please do less nesting?

```suggestion
if (functionResponse?.id) {
  toolCallId = functionResponse.id;
}
if (functionResponse?.response) {
  toolResponse = functionResponse.response;
}
```